### ISSUE TITLE
DatePicker no longer expands partially entered dates

### ISSUE THREAD

**Issue by guttormvik2** [vaadin/flow-components#6065]
### Describe the bug

When I write "1902" into a DatePicker, I expect it to expand it to "19.02.2024", since I've defined multiple formats in DatePickerI18n.setDateFormats.
Currently, it does *nothing*. No ValueChanged. No validation error

### Expected-behavior

DatePicker should loop over the formats provided in DatePickerI18n.setDateFormats, as specified in its javadoc
Any parts that are not provided should be taken from todays date.
Given:
* today is 19.02.2024:
* setDateFormats("dd.MM.yyyy", "dd", "ddMM", "ddMMyy", "ddMMyyyy", "yyyy-MM-dd", "MM/dd/yyyy")

The following expansions should happen
* 01 -> 01.02.2024
* 0101 -> 01.01.2024
* 010125 -> 01.01.2025
* 02/01 -> 01.02.2024
* etc

### Reproduction

```var someDate = new DatePicker("Some date");
var i18n = new DatePickerI18n()
    .setDateFormats
        ( "dd.MM.yyyy"
        , "dd"
        , "ddMM"
        , "ddMMyy"
        , "ddMMyyyy"
        , "yyyy-MM-dd"
        , "MM/dd/yyyy"
        );
someDate.setI18n(i18n);
someDate.setAutoOpen(false);
someDate.addValueChangeListener(event -> {
    Notification.show("Date changed: " + event.getValue());
});
add(someDate);
```

### System Info

Vaadin 24.3.5
Firefox 122.0.1 / Edge 121.0.2277.128
Windows 10 22H2

**Comment by guttormvik2** [comment]
@sissbruecker, why no longer a bug?

**Comment by sissbruecker** [comment]
We assign labels ourselves after triaging the issue. Also not sure if we explicitly support patterns like `ddMM`, which look like they might cause this issue.

**Comment by sissbruecker** [comment]
Can confirm it's a bug, it's a regression from https://github.com/vaadin/flow-components/pull/5752, since v24.3.

This function tries to create a short year format from a configured format, but returns undefined for a lot of cases:
https://github.com/vaadin/flow-components/blob/f05d10386e99a45f42fa77a4235d49911818de11/vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js#L58-L66

Then the parse function just uses the result of that function without checking its result for undefined. This leads to the exception when trying to parse the date with date-fns:
https://github.com/vaadin/flow-components/blob/f05d10386e99a45f42fa77a4235d49911818de11/vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js#L132-L141

**Comment by guttormvik2** [comment]
@sissbruecker, I thought that might be the case, so no problem. I just didn't see it when I I looked at a couple of other issues.

The date expansion is a result of a complaint we raised a couple of years ago; We wanted access to the failed text on the serverside to do this kind of expansion, but vaadin gave us this mechanism instead.

I see that the javadoc on setDateFormats doesn't explicitly mention expansion, but here is an excerpt of the discussion we had back in 2021:

> Rolf Smeds
Ok so those general parsing and formatting needs would indeed be solved by the feature we're currently planning: The idea is to use date-fns for parsing and formatting, and you would be able to provide multiple parsing formats, so configuring the datepicker with setParsers("d", "dd", "ddMM", "ddMMyy", "ddMMyyyy"); and setFormat("dd.MM.yyyy"); would give you exactly that.

This worked "until recently", but since our vaadin20+ version is in development, no-one is properly using the app, so I don't know when it stopped working. I've tried reverting to 24.3.0 but it didn't work there either
